www/src/content/blog/2025-05-13-An-introduction-to-rate-limiting.mdx

```diff
@@ -0,0 +1,202 @@
+
+### User IDs
+
+By far the most accurate and secure identifier is a unique user or account ID, typically passed through authentication headers. A unique ID ensures that requests are counted against a specific account, making it possible to enforce tiered limits for different subscriptions.
+
+Account or user IDs are the most broadly useful identifier, as they can be used to meet an app’s business needs, while also providing some protection against resource exhaustion and malicious request patterns. 
+
+Note that while an ID uniquely identifies an *account*, it doesn’t inherently reveal anything about the user behind it or their location, making it a solid choice from an ethical perspective as well.
+
+### Request IPs
+
+Many examples use the client IP, as it’s relationship to requests is fairly intuitive, and the implementation doesn’t rely on the app’s business logic (e.g., for a user ID). IP limits are sub-optimal for capacity management though, as they do not necessarily correspond to unique users.
+
+Multiple users could share the same IP, in which case one user could exhaust the limit for all. Conversely, a single client could also circumvent limits by accessing services using different IPs by using a VPN or a proxy.
+
+While IPs should not be used as a primary limiter, they play a vital role in security-oriented rate limiting, especially in auth flows. If a user hasn’t yet been authenticated (or even registered), IP limits can prevent a malicious client from bombarding auth services, either to overwhelm them or in an effort to brute-force access. In more sophisticated rate limiting implementations, IPs with a pattern of malicious requests can also be blacklisted.
+
+### Geolocation
+
+Location-based limiting is far less common outside of enterprise-scale apps, perhaps in part because it is comparatively resource intensive. It can be used to dynamically manage limits by region based on expected traffic patterns. If request count decreases at night, for example, the rate limit for a given region could be lowered until the next morning.
+
+This type of limiting is especially useful in a security context, as it can mitigate (and flag) suspicious request patterns. Doing so effectively requires some kind of geo-fencing though, as well as a service dedicated to managing the dynamic limits.
+
+### Resource Limiting
+
+Of course, it’s also possible to limit how much a resource is used at all, whether it’s a server, an  endpoint, or a single workflow. Since resource limiters don’t discriminate between clients, they’re not recommended for most rate limiting use-cases. They’re mostly useful for capacity management, and serve as a last resort against unexpected floods of traffic.
```

Comment:
```suggestion
Of course, it’s also possible to limit how much a resource is used at all, whether it’s a server, an endpoint, or a single workflow. Since resource limiters don’t discriminate between clients, they’re not recommended for most rate limiting use-cases. They’re mostly useful for capacity management, and serve as a last resort against unexpected floods of traffic.
```